milabench/executors.py

```diff
@@ -0,0 +1,612 @@
+    """Execute a command
+
+    Arguments:
+        pack: `BasePackage`'s instance from which `execute()` will be used to
+              perform the command
+        *cmd_argv: command line arguments list to execute
+        **kwargs: kwargs to be passed to the `pack.execute()`
+    """
+    def __init__(
+            self,
+            pack:pack.BasePackage,
+            *cmd_argv,
+            **kwargs
+    ) -> None:
+        if isinstance(pack, Executor):
+            raise ValueError(f"{self.__class__.__name__} does not accept nested"
+                             f" {Executor.__class__.__name__}")
+        super().__init__(
+            pack,
+            **kwargs
+        )
+        self.cmd_argv = cmd_argv
+
+    def _argv(self, **kwargs) -> List:
+        del kwargs
+        return [*self.cmd_argv]
+
+
+class PackExecutor(CmdExecutor):
+    """Execute a `Package`'s script. If not specified, the `Package`'s
+    main_script will be used
+
+    Arguments:
+        pack: `Package`'s instance from which `execute()` will be used to
+              perform the command
+        *script_argv: script's command line arguments list. If the first
+                      argument is a file or directory that can be found, the
+                      file will be used instead of the `pack`'s main_script
+        **kwargs: kwargs to be passed to the `pack.execute()`
+    """
+    def __init__(
+            self,
+            pack:pack.Package,
+            *script_argv,
+            **kwargs
+    ) -> None:
+        script = script_argv[:1]
+        if script and XPath(script[0]).exists():
+            script = script[0]
+            script_argv = script_argv[1:]
+        else:
+            script = None
+
+        super().__init__(pack, *script_argv, **kwargs)
+
+        self.script = script
+
+    def _argv(self, **kwargs) -> List:
+        main = self.script or self.pack.main_script
+        if not XPath(main).is_absolute():
+            abs_main = self.pack.dirs.code / main
+        else:
+            abs_main = main
+
+        if not abs_main.exists():
+            raise FileNotFoundError(
+                f"Cannot run script or directory because it does not exist: {main}"
+            )
+
+        if abs_main.is_dir():
+            main = ["-m", str(main)]
+        else:
+            main = [str(abs_main)]
+        return main + super()._argv(**kwargs)
+
+
+class VoidExecutor(CmdExecutor):
+    """Execute nothing"""
+    def __init__(
+            self,
+            pack:pack.BasePackage,
+            *argv,
+            **kwargs
+    ) -> None:
+        super().__init__(pack, "true", *argv, **kwargs)
+
+
+class TorchRun(Executor):
+    def __init__(self, executor, **kwargs) -> None:
+        super().__init__(executor, **kwargs)
+
+    def _argv(self, **kwargs):
+        devices = self.pack.config.get("devices", [])
+        nproc = len(devices)
+        if nproc > 1:
+            return ["torchrun", f"--nproc_per_node={nproc}", "--", "-m"]
```

Comment:
What if the sub-executor puts a script path and not module name after the `-m`, though.

Reply:
should be good now